app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,159 @@
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.APP_ICON
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.ERROR
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.CUSTOM_TAB
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.LINK
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.UNKNOWN
+
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.HOT
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false]
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppAllStartup] and finally report them on Activity's onResume() function.
+ */
+@Suppress("TooManyFunctions")
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    private var onCreateData: AppAllStartup? = null
+    private var onRestartData: Pair<Type, Boolean>? = null
+    private var onNewIntentData: Source? = null
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, false),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, true),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    /**
+     * we are not setting [Source] in this method since source is derived from a [SafeIntent].
+     * therefore source either gets set in oncreate() or onNewIntent().
```

Comment:
This is false: we're only looking for a source from `onNewIntent`, right?

Also, I'd put this comment in the function body  for two reasons:
- it's closer to the code it affects so if the implementation changes, the comment is more likely to be updated
- kdoc is generally reserved for information relevant to callers, not implementation details

Reply:
When its a cold or warm startup , we are getting the source from onCreate (using the launching intent). Whenever onCreate is called, onNewIntent() is not called.